### PR TITLE
VReplication: Ensure that RowStreamer uses optimal index when possible

### PR DESCRIPTION
**Pull request by mattlord**
## Description

In this PR we add a `force index` clause (and not always `primary`) when we have PK or equivalent columns covered by an existing index, and we then use all of those columns in the `ORDER BY` clause. So it’s for the low level `rowstreamer` source query we apply at the `mysqld` layer to get rows for streaming, before applying the vreplication filters in the `vstreamer` layer in the source tablet and finally sending the matching rows on to the `vcopier` in the target tablet. Using an index for the `ORDER BY` clause there in the `mysqld` layer is what allows us to immediately start streaming rows rather than having to read every matching row and sort them all on disk with a filesort in `mysqld` before the `vstreamer` gets any rows back from `mysqld`/`rowstreamer`.

> **Note**
> ✅ My current plan is to hold off on merging this until after the upcoming 18.0-RC cutoff (scheduled for Sept 29th) so that it has the full v19 dev cycle to bake and surface any unforeseen edge cases.

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/11991
  - Fixes: https://github.com/vitessio/vitess/issues/13919

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by rohit-nayak-ps — APPROVED** [review]
lgtm

File: go/vt/vttablet/tabletserver/vstreamer/rowstreamer.go [hunk 1]
```diff
@@ -263,18 +264,27 @@ func (rs *rowStreamer) buildSelect() (string, error) {
...
 	}
-	buf.Myprintf(" from %v", sqlparser.NewIdentifierCS(rs.plan.Table.Name))
+	// If we know the index name that we should be using then tell MySQL
+	// to use it if possible. This helps to ensure that we are able to
+	// leverage the ordering from the index itself and avoid having to
+	// do a FILESORT of all the results. This index should contain all
+	// of the PK columns which are used in the ORDER BY clause below.
+	var indexHint string
+	if st.PKIndexName != "" {
+		indexHint = fmt.Sprintf(" force index (`%s`)", st.PKIndexName)
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletserver/vstreamer/rowstreamer.go:274]: We might want to escape the index name properly before putting it into `fmt.Sprintf`. Its name just might have backticks.

**Review by shlomi-noach — APPROVED** [review]
Makes perfect sense, with one minor comment.

File: go/vt/vttablet/tabletserver/vstreamer/rowstreamer.go (hunk 1 above)
Comment (mattlord) [reply go/vt/vttablet/tabletserver/vstreamer/rowstreamer.go:274]: Will do. Thanks!

**Pull request merged**